lib/service/modules/dataStore.js

```diff
@@ -0,0 +1,183 @@
+function initDataStore(appMetadata, requestContext) {
+  function DataStore(storeOptions = {}) {
+    this.appMetadata = appMetadata;
+    this.requestContext = requestContext;
+    const skipBl = this.skipBl = storeOptions.skipBl || false;
+    const useMasterSecret = this.useMasterSecret = storeOptions.useMasterSecret || false;
+
+    function collection(collectionName) {
+      function _buildKinveyRequest() {
+        const url = `${appMetadata.baasUrl}/appdata/${appMetadata._id}/${collectionName}/`;
+        const headers = {
+          'Content-Type': 'application/json',
+          'X-Kinvey-API-Version': requestContext.apiVersion,
+        };
+
+        if (skipBl) {
+          headers['x-kinvey-skip-business-logic'] = true;
+        }
+
+        const requestOptions = {
+          url,
+          headers,
+          json: true
+        };
+
+        if (storeOptions.useMasterSecret) {
+          requestOptions.auth = {
+            user: appMetadata._id,
+            pass: appMetadata.mastersecret
+          };
+        } else {
+          requestOptions.headers.authorization = requestContext.authorization;
+        }
+
+        return requestOptions;
+      }
+
+      function _makeRequest(requestOptions, callback) {
+        request(requestOptions, (err, res, body) => {
+          if (err) {
+            return callback(err);
+          } else if (res.statusCode > 299) {
+            const error = new Error(body.error);
+            error.description = body.description;
+            error.debug = body.debug;
+            return callback(err);
+          }
+          return callback(null, body);
+        });
+      }
+
+      function _generateQueryString(query) {
+        return query.toQueryString();
+      }
+
+      function save(entity, callback) {
+        const requestOptions = _buildKinveyRequest();
+
+        if (entity._id) {
+          requestOptions.method = 'PUT';
+          requestOptions.url += entity._id;
+        } else {
+          requestOptions.method = 'POST';
+        }
+
+        requestOptions.body = entity;
+
+        _makeRequest(requestOptions, callback);
+      }
+
+      function findById(entityId, callback) {
+        const requestOptions = _buildKinveyRequest();
+
+        requestOptions.method = 'GET';
+        requestOptions.url += entityId;
+        _makeRequest(requestOptions, callback);
+      }
+
+      function find(query, callback) {
+        const requestOptions = _buildKinveyRequest();
+
+        if (typeof query === 'function' && !callback) {
+          callback = query;
+          query = null;
+        }
+
+        requestOptions.method = 'GET';
+
+        if (query) {
+          requestOptions.qs = _generateQueryString(query);
+        }
+
+        _makeRequest(requestOptions, callback);
+      }
+
+      function removeById(entityId, callback) {
+        const requestOptions = _buildKinveyRequest();
+
+        requestOptions.method = 'DELETE';
+        requestOptions.url += entityId;
```

Comment:
Should we add some protection here (and in similar occurrences) from null/undefined `entityId` params?